swarm/swarm.go

```diff
@@ -0,0 +1,382 @@
+			}
+		case b := <-s.incoming:
+			s.metrics.IncCounter(metricsPrefix + "incoming.all")
+			m, err := decodeMessage(b)
+			if err != nil {
+				log.Errorf("SWARM: Failed to decode message: %v", err)
+			} else if m.Type == sharedValue {
+				s.metrics.IncCounter(metricsPrefix + "incoming.shared")
+				log.Debugf("SWARM: %s got shared value from %s: %s: %v", s.Local().Name, m.Source, m.Key, m.Value)
+				s.shared.set(m.Source, m.Key, m.Value)
+			} else if m.Type == broadcast {
+				s.metrics.IncCounter(metricsPrefix + "incoming.broadcast")
+				log.Debugf("SWARM: got broadcast value: %s %s: %v", m.Source, m.Key, m.Value)
+				for k, l := range s.listeners {
+					if k == m.Key {
+						// assuming buffered listener channels
+						select {
+						case l <- &Message{
+							Source: m.Source,
+							Value:  m.Value,
+						}:
+						default:
+						}
+					}
+				}
+			}
+		case req := <-s.getValues:
+			log.Debugf("SWARM: getValues for key: %s", req.key)
+			req.ret <- s.shared[req.key]
+		case <-s.leave:
+			log.Debugf("SWARM: %s got leave signal", s.Local())
+			if s.mlist == nil {
+				log.Warningf("SWARM: Leave called, but %s already seem to be left", s.Local())
+				return
+			}
+			if err := s.mlist.Leave(s.leaveTimeout); err != nil {
+				log.Errorf("SWARM: Failed to leave mlist: %v", err)
+			}
+			if err := s.mlist.Shutdown(); err != nil {
+				log.Errorf("SWARM: Failed to shutdown mlist: %v", err)
+			}
+			log.Infof("SWARM: %s left", s.Local())
+			return
+		}
+	}
+}
+
+// Local is a getter to the local member of a swarm.
+func (s *Swarm) Local() *NodeInfo {
+	if s == nil {
+		log.Errorf("swarm is nil")
+		return nil
+	}
+	if s.mlist == nil {
+		log.Warningf("deprecated way of getting local node")
+		return s.local
+	}
+	mlNode := s.mlist.LocalNode()
+	return &NodeInfo{
+		Name: mlNode.Name,
+		Addr: mlNode.Addr,
+		Port: mlNode.Port,
+	}
+}
+
+func (s *Swarm) broadcast(m *message) error {
+	m.Source = s.Local().Name
+	b, err := encodeMessage(m)
+	if err != nil {
+		return err
+	}
+
+	s.outgoing <- &outgoingMessage{
+		message: m,
+		encoded: b,
+	}
+	return nil
+}
+
+// Broadcast sends a broadcast message with a value to all peers.
+func (s *Swarm) Broadcast(m interface{}) error {
+	return s.broadcast(&message{Type: broadcast, Value: m})
+}
+
+// ShareValue sends a broadcast message with a sharedValue to all
+// peers. It implements the ratelimit.Swarmer interface.
+func (s *Swarm) ShareValue(key string, value interface{}) error {
+	return s.broadcast(&message{Type: sharedValue, Key: key, Value: value})
+}
+
+// Values sends a request and wait blocking for a response. It
+// implements the ratelimit.Swarmer interface.
+func (s *Swarm) Values(key string) map[string]interface{} {
+	req := &valueReq{
+		key: key,
+		ret: make(chan map[string]interface{}),
+	}
+	s.getValues <- req
+	return <-req.ret
+}
+
+// Leave sends a signal for the local node to leave the Swarm.
+func (s *Swarm) Leave() {
+	close(s.leave)
+	s.cleanupF()
```

Comment:
i can't find it, is it sure that s.cleanupF is not nil? I would make sure to set a default noop in Join() like `if cleanupF == nil { cleanupF = func() {} }`